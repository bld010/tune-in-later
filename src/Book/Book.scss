


.Book {
  margin: 15px; 
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
/*   border: 1px solid black; */
  background: white;
  width: 280px;
  height: 160px;
  border-radius: 8px;
  margin-top: 45px;
  box-shadow: 1px 1px 5px 5px lightgray;

  & div.img {
  height: 110px;
  width: 80px;
  margin: 20px;
  background: lightblue;
  
  }

  &.favorited {
    background-color: yellow;
  }

  & .bookInfo {
    width: 200px;
  }

  & h2 {
    margin-bottom: 0px;
/*   padding: 0; */
  }

  & h3 {
    margin-top: 0px;
  }

  & h4 {
    margin-top: 35px;
  }


  & button {
    border: none;
    color: red;
    position: relative;
    left: 130px;
    top: -10px;
  }

  & button:hover {
    cursor: pointer;
  }

}














