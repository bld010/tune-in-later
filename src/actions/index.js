export const setCurrentUser = user => ({
  type: 'SET_CURRENT_USER',
  user
});

export const setCurrentUserFavorites = favorites => ({
  type: 'SET_CURRENT_USER_FAVORITES',
  favorites
});

export const displaySearchResults = searchResults => ({
  type: 'DISPLAY_SEARCH_RESULTS',
  searchResults

})

export const addUserFavorite = favorite => ({
  type: 'ADD_USER_FAVORITE',
  favorite
})