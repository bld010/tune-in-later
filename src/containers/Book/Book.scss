.Book {
  width: 400px;
  min-width: 465px;
  max-height: 200px;
  margin: 25px 10px 25px 35px; 
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  background: white;
  border-radius: 8px;
  box-shadow: 1px 1px 5px 5px lightgray;

  & div.img img{
  min-height: 140px;
  min-width: 140px;
  margin: 20px;
  border-radius: 8px;
  border: 1px solid black;
  }

  &.favorited {
    background-color: #F7E1DA;
  }

  & .bookInfo {
    text-align: center;
    display: flex;
    flex-direction: column;
    height: 60%;
    margin: 25px 0px 0px auto;
  }

  & h2 {
    margin: 25px 0px 0px 0px;
  }

  & h3 {
    margin: 25px 0px 0px 0px;
    font-style: italic;
    color: #666888;
  }

  & h4 {
    margin-top: 10px;
  }


  & button {
    border: none;
    background-color: transparent;
    color: #666888;
    position: relative;
    margin: 0px 2px 0px 0px;
    top: 115px;
    height: 30px;
    width: 75px;

    & img {
      height: 45px;
      width: 45px;
      margin: 0px 0px 5px 0px;
    }
  }

  & button:hover {
    cursor: pointer;
  }

}














