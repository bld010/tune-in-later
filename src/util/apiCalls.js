export const landingFetch = () => {
  const romanceFetch = fetch('https://itunes.apple.com/search?media=audiobook&term=romance');
  const fantasyFetch = fetch('https://itunes.apple.com/search?media=audiobook&term=fantasy');
  const biographyFetch = fetch('https://itunes.apple.com/search?media=audiobook&term=biography');
  const historyFetch = fetch('https://itunes.apple.com/search?media=audiobook&term=history');
  const horrorFetch = fetch('https://itunes.apple.com/search?media=audiobook&term=horror');

  return Promise.all([romanceFetch, fantasyFetch, biographyFetch, historyFetch, horrorFetch])
    .then(responses => Promise.all(responses.map(response => response.json())))
    .then(data => {
      return data.map((datum, i) => {
        return datum.results.map(obj => {
          switch (i) {
            case 0:
              return { ...obj, filterType: 'romance' };
            case 1:
              return { ...obj, filterType: 'fantasy' };
            case 2:
              return { ...obj, filterType: 'biography' };
            case 3:
              return { ...obj, filterType: 'history' };
            case 4:
              return { ...obj, filterType: 'horror' };
            default:
              return obj;
          }
          })
        })
      })
    .catch(err => console.log(err, 'error in apiCalls landingFetch'))

}

export const authorFetch = (authorFirstName, authorLastName) => {
  return fetch(`https://itunes.apple.com/search?media=audiobook&term=${authorFirstName+authorLastName}`)
    .then(response => response.json())
    .catch(err => console.log(err, 'error in apiCalls authorFetch'))

}

export const loginUser = (email, password) => {
  const user = {
    "email": email,
    "password": password
  };

  const options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  }

  return fetch('http://localhost:3001/api/v1/login/', options)    
}

export const addFavoritesToApi = (book, userId) => {
  const favoriteBook = {
    "book_id": book.id,
    "author_name": book.artist,
    "book_name": book.title,
     "artwork_url": book.image,
    "release_date": book.date,
    "description": book.description
  }
  const options = {
    method: "POST",
    body: JSON.stringify(favoriteBook),
    headers: {
      "Content-Type": "application/json"
    }
  }
  return fetch(`http://localhost:3001/api/v1/users/${userId}/bookfavorites`, options)
}