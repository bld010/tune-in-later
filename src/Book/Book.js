import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addUserFavorite, setCurrentUserFavorites, setSelectedBook, deleteUserFavorite } from '../actions';
import { addFavoriteToApi, deleteFavoriteFromApi } from '../util/apiCalls';
import star from '../images/star.svg';
import activeStar from '../images/active-star.svg';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import './Book.scss';

class Book extends Component {
  state={
    buttonWasClicked: false
  }

  handleFavorite = (e) => {
    e.stopPropagation();
    e.preventDefault();
    if (!this.props.favorites.find(obj => obj.book_id === this.props.book.id)) {
      addFavoriteToApi(this.props.book, this.props.currentUser.id)
        .then(response => response.json())
        .then(data => this.props.addUserFavorite(data, this.props.currentUser.id))
        .then(data => console.log('added book', data))
        .catch(error => console.error(error));
    } else {
      deleteFavoriteFromApi(this.props.book, this.props.currentUser.id)
        .then(resp => {
          if(!resp.ok) {
          throw Error('There was an error deleting the favorite')
        }
          this.props.deleteUserFavorite(this.props.book)
        })
        .catch(error => console.error(error.message));
    } 
  }

  handleButtonClick = (e) => {
    e.stopPropagation();
    e.preventDefault();
    this.setState({buttonWasClicked: true});
    if (this.props.currentUser) {
      this.handleFavorite(e)
      this.setState({buttonWasClicked: false})
    }
  }

  
  render() {
    const { title, artist, image } = this.props.book;
    const isFavorite = this.props.favorites.find(obj => obj.book_id === this.props.book.id);
    const favoriteClass = isFavorite ? 'Book favorited' : 'Book'; 
    const favoriteStar = isFavorite ? activeStar : star;
    const bookElement = 
    <article className={favoriteClass} onClick={() => this.props.setSelectedBook(this.props.book)}>
      <div className="img">
        <img src={image} alt='' />
      </div>
      <div className="bookInfo">
        <h2>{title}</h2>
        <h3>{artist}</h3>
      </div>
      <button onClick={this.handleButtonClick}><img src={favoriteStar} alt=''/>Favorite</button>
    </article>;

    if (!this.props.currentUser && this.state.buttonWasClicked) {
      window.scrollTo(0,0);
      return (
        <>
          {bookElement}
          <Redirect to='/login' />
        </>
      )
    }
    return(
      <>
      {bookElement}
      </>
    )
  }

}

const mapStateToProps = state => ({
  currentUser: state.currentUser,
  favorites: state.favorites
})

const mapDispatchToProps = dispatch => ({
  setSelectedBook: book => dispatch(setSelectedBook(book)),
  setCurrentUserFavorites: favorites => dispatch(setCurrentUserFavorites(favorites)),
  addUserFavorite: (favorite, id) => dispatch(addUserFavorite(favorite, id)),
  deleteUserFavorite: (favorite) => dispatch(deleteUserFavorite(favorite))
})

export default connect(mapStateToProps, mapDispatchToProps)(Book);

Book.propTypes = {
  currentUser: PropTypes.object,
  favorites: PropTypes.array.isRequired,
  setSelectedBook: PropTypes.func.isRequired,
  setCurrentUserFavorites: PropTypes.func.isRequired,
  addUserFavorite: PropTypes.func.isRequired,
  deleteUserFavorite: PropTypes.func.isRequired,
}
