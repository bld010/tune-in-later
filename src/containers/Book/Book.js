import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addUserFavorite, setCurrentUserFavorites, setSelectedBook, deleteUserFavorite } from '../../actions';
import { addFavoriteToApi, deleteFavoriteFromApi } from '../../util/apiCalls';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import './Book.scss';

export class Book extends Component {

  state={
    buttonWasClicked: false
  }


  handleFavorite = (e) => {
    e.stopPropagation();
    e.preventDefault();
    if (!this.props.favorites.find(obj => obj.book_id === this.props.book.id)) {
      addFavoriteToApi(this.props.book, this.props.currentUser.id)
        .then(response => response.json())
        .then(data => this.props.addUserFavorite(data, this.props.currentUser.id))
        .then(data => console.log('added book', data))
        .catch(error => console.error(error));
    } else {
      deleteFavoriteFromApi(this.props.book, this.props.currentUser.id)
        .then(resp => {
          if(!resp.ok) {
          throw Error('There was an error deleting the favorite')
        }
          this.props.deleteUserFavorite(this.props.book)
        })
        .catch(error => console.error(error.message));
    } 
  }

  handleButtonClick = (e) => {
    e.stopPropagation();
    e.preventDefault();
    this.setState({buttonWasClicked: true});
    if (this.props.currentUser) {
      this.handleFavorite(e)
      this.setState({buttonWasClicked: false})
    }
  }

  
  render() {

    // console.log(this.props.selectedBook)
    const { title, artist, filterType, price, image } = this.props.book;
    const isFavorite = this.props.favorites.find(obj => obj.book_id === this.props.book.id);
    const favoriteClass = isFavorite ? 'Book favorited' : 'Book'; 
    const bookElement = <article className={favoriteClass} onClick={() => this.props.setSelectedBook(this.props.book)}>
    <div className="img">
      <img src={image} alt='' />
    </div>
    <div className="bookInfo">
      <h2>{title}</h2>
      <h3>{artist}</h3>
      <button onClick={this.handleButtonClick}>Favorite</button>
    </div>
  </article>;




    if (!this.props.currentUser && this.state.buttonWasClicked) {
      window.scrollTo(0,0);
      return (
        <>
          {bookElement}
          <Redirect to='/login' />
        </>
      )
    }
    return(
      <>
      {bookElement}
      </>
    )
  }
}

const mapStateToProps = state => ({
  currentUser: state.currentUser,
  favorites: state.favorites
})

const mapDispatchToProps = dispatch => ({
  setSelectedBook: book => dispatch(setSelectedBook(book)),
  setCurrentUserFavorites: favorites => dispatch(setCurrentUserFavorites(favorites)),
  addUserFavorite: (favorite, id) => dispatch(addUserFavorite(favorite, id)),
  deleteUserFavorite: (favorite) => dispatch(deleteUserFavorite(favorite))
})

export default connect(mapStateToProps, mapDispatchToProps)(Book);

Book.propTypes = {
  book: PropTypes.object.isRequired,
  currentUser: PropTypes.object,
  favorites: PropTypes.array.isRequired,
  setSelectedBook: PropTypes.func.isRequired,
  setCurrentUserFavorites: PropTypes.func.isRequired,
  addUserFavorite: PropTypes.func.isRequired,
  deleteUserFavorite: PropTypes.func.isRequired,
}
