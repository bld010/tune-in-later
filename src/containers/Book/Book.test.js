import React from 'react';
import { Book, mapStateToProps, mapDispatchToProps } from './Book';
import { setSelectedBook, addUserFavorite, deleteUserFavorite } from '../../actions';
import {addFavoriteToApi, deleteFavoriteFromApi} from '../../util/apiCalls';
import { shallow } from 'enzyme';

jest.mock('../../util/apiCalls');

describe('Book', () => {
  let wrapper, mockDispatch, mappedProps;
  const mockBookProp = { id: 1, title: 'Book', artist: 'David Gitlen', image: 'https://davidgitlen.com/author.png' };
  const mockSetSelectedBook = jest.fn();
  const mockSetCurrentUserFavorites = jest.fn();
  const mockAddUserFavorite = jest.fn();
  const mockDeleteUserFavorite = jest.fn();

  beforeEach(() => {
    addFavoriteToApi.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => {
          return Promise.resolve(mockBookProp)
        }
      });
    });
    deleteFavoriteFromApi.mockImplementation(() => {
      return Promise.resolve({
        ok: true
      });
    });
    mockDispatch = jest.fn();
    mappedProps = mapDispatchToProps(mockDispatch);
    wrapper=shallow(
      <Book 
        book={mockBookProp}
        favorites={[]}
        currentUser={null}
        setSelectedBook={mockSetSelectedBook}
        setCurrentUserFavorites={mockSetCurrentUserFavorites}
        addUserFavorite={mockAddUserFavorite}
        deleteUserFavorite={mockDeleteUserFavorite}
      />
    )
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if the book has been favorited', () => {
    const wrapper = shallow(<Book
      book={mockBookProp}
      favorites={[{book_id:1}]}
      currentUser={null}
      setSelectedBook={mockSetSelectedBook}
      setCurrentUserFavorites={mockSetCurrentUserFavorites}
      addUserFavorite={mockAddUserFavorite}
      deleteUserFavorite={mockDeleteUserFavorite}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should fire handleButtonClick when the favorites button is clicked', () => {
    const mockEvent = {
      stopPropagation: jest.fn(),
      preventDefault: jest.fn()
    }
    wrapper.instance().handleButtonClick = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('button').simulate('click', mockEvent);

    expect(wrapper.instance().handleButtonClick).toHaveBeenCalledTimes(1);
  });

  describe('handleButtonClick', () => {

    it('should update state appropriately when a click is detected', () => {

      const wrapper = shallow(<Book
        book={mockBookProp}
        favorites={[{ book_id: 1 }]}
        currentUser={{ id: 2, name: 'Divad' }}
        setSelectedBook={mockSetSelectedBook}
        setCurrentUserFavorites={mockSetCurrentUserFavorites}
        addUserFavorite={mockAddUserFavorite}
        deleteUserFavorite={mockDeleteUserFavorite}
      />);

      const mockEvent = {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn()
      };

      const initialState = {buttonWasClicked: true};

      wrapper.instance().setState(initialState);

      expect(wrapper.instance().state).toEqual({buttonWasClicked: true});

      wrapper.instance().handleButtonClick(mockEvent);

      expect(wrapper.instance().state).toEqual({buttonWasClicked: false});


    });
  });

  describe('handleFavorite', () => {

    it('should be called when handleButtonClick is fired and a user is logged in', () => {
      const wrapper = shallow(<Book
        book={mockBookProp}
        favorites={[{ book_id: 1 }]}
        currentUser={{id: 2, name: 'Divad'}}
        setSelectedBook={mockSetSelectedBook}
        setCurrentUserFavorites={mockSetCurrentUserFavorites}
        addUserFavorite={mockAddUserFavorite}
        deleteUserFavorite={mockDeleteUserFavorite}
      />);

      const mockEvent = {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn()
      };

      wrapper.instance().handleFavorite = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.instance().handleButtonClick(mockEvent);

      expect(wrapper.instance().handleFavorite).toHaveBeenCalledWith(mockEvent);
    });

    it('should fire addFavoriteToApi with the current book and user id passed in if handleButtonClick is called and the book is not currently a favorite', () => {
      const mockEvent = {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn()
      }

      const wrapper = shallow(
        <Book
          book={mockBookProp}
          favorites={[{ book_id: 20 }, { book_id: 30 }]}
          currentUser={{ name: 'Brianna', id: 12 }}
          setSelectedBook={mockSetSelectedBook}
          setCurrentUserFavorites={mockSetCurrentUserFavorites}
          addUserFavorite={mockAddUserFavorite}
          deleteUserFavorite={mockDeleteUserFavorite}
        />
      )
      wrapper.instance().handleButtonClick(mockEvent);
      expect(addFavoriteToApi).toHaveBeenCalledWith(mockBookProp, 12);
    });

    it('should fire deleteFavoriteFromApi with the current book and user id passed in if handleButtonClick is called and the book is currently a favorite', () => {
      const mockEvent = {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn()
      }

      const wrapper = shallow(
        <Book
          book={mockBookProp}
          favorites={[{ book_id: 1 }, { book_id: 30 }]}
          currentUser={{ name: 'Brianna', id: 12 }}
          setSelectedBook={mockSetSelectedBook}
          setCurrentUserFavorites={mockSetCurrentUserFavorites}
          addUserFavorite={mockAddUserFavorite}
          deleteUserFavorite={mockDeleteUserFavorite}
        />
      );
      wrapper.instance().handleButtonClick(mockEvent);
      expect(deleteFavoriteFromApi).toHaveBeenCalledWith(mockBookProp, 12);
    });

  });

  describe('mapStateToProps', () => {

    it('should return an object with appropriate data', () => {

      const mockState = {
        currentUser: {id: 1, name: 'Divad', email: 'divad@turing.io'},
        favorites: [{id: 20}, {id: 30}],
        searchResults: [{id: 10}]
      }

      const expected = {
        currentUser: { id: 1, name: 'Divad', email: 'divad@turing.io' },
        favorites: [{ id: 20 }, { id: 30 }]
      }

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {

    it('calls dispatch with action object from setSelectedBook when card is clicked', () => {
      const actionToDispatch = setSelectedBook(mockBookProp);

      mappedProps.setSelectedBook(mockBookProp);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with action object from addUserFavorite when addFavoriteToApi fetch resolves', () => {
      const actionToDispatch = addUserFavorite(mockBookProp);

      mappedProps.addUserFavorite(mockBookProp);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with action object from deleteUserFavorite when deleteFavoriteFromApi fetch resolves', () => {
      const actionToDispatch = deleteUserFavorite(mockBookProp);

      mappedProps.deleteUserFavorite(mockBookProp);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});