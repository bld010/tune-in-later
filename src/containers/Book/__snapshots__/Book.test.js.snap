// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book should match the snapshot with all data passed in correctly 1`] = `
<Fragment>
  <article
    className="Book"
    onClick={[Function]}
  >
    <div
      className="img"
    >
      <img
        alt=""
      />
    </div>
    <div
      className="bookInfo"
    >
      <h2>
        Book
      </h2>
      <h3 />
    </div>
    <button
      onClick={[Function]}
    >
      <img
        alt=""
        src="star.svg"
      />
      Favorite
    </button>
  </article>
</Fragment>
`;
